src/runtime/micro/openocd_low_level_device.cc

```diff
@@ -0,0 +1,248 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file openocd_low_level_device.cc
+ */
+#include <sstream>
+#include <iomanip>
+
+#include "micro_common.h"
+#include "low_level_device.h"
+#include "tcl_socket.h"
+
+namespace tvm {
+namespace runtime {
+
+/*!
+ * \brief OpenOCD low-level device for uTVM micro devices connected over JTAG
+ */
+class OpenOCDLowLevelDevice final : public LowLevelDevice {
+ public:
+  /*!
+   * \brief constructor to initialize connection to openocd device
+   * \param base_addr base address of the device
+   * \param server_addr address of the OpenOCD server to connect to
+   * \param port port of the OpenOCD server to connect to
+   */
+  explicit OpenOCDLowLevelDevice(std::uintptr_t base_addr,
+                                 const std::string& server_addr,
+                                 int port) : socket_() {
+      socket_.Connect(tvm::common::SockAddr(server_addr.c_str(), port));
+      socket_.SendCommand("reset halt");
+      base_addr_ = base_addr;
+      CHECK(base_addr_ % 8 == 0) << "base address not aligned to 8 bytes";
+  }
+
+  void Read(DevBaseOffset offset, void* buf, size_t num_bytes) {
+    if (num_bytes == 0) {
+      return;
+    }
+
+    // TODO(weberlo): Refactor between read and write.
+    // Check if we need to chunk this write request.
+    if (num_bytes > kMemTransferLimit) {
+      DevBaseOffset curr_offset = offset;
+      char* curr_buf_ptr = reinterpret_cast<char*>(buf);
+      while (num_bytes != 0) {
+        size_t amount_to_read;
+        if (num_bytes > kMemTransferLimit) {
+          amount_to_read = kMemTransferLimit;
+        } else {
+          amount_to_read = num_bytes;
+        }
+        Read(offset, reinterpret_cast<void*>(curr_buf_ptr), amount_to_read);
+        offset += amount_to_read;
+        curr_buf_ptr += amount_to_read;
+        num_bytes -= amount_to_read;
+      }
+      return;
+    }
+    {
+      socket_.SendCommand("array unset output");
```

Comment:
Given that we have multiple places where we need to construct the command, how about change the API to

```
socket_.cmd_stream() << "array unset output";
socket_.SendCommand();
```